File: rust/src/hash.rs
//! Deterministic, non-cryptographic content hashing.
//!
//! Cairn uses the 64-bit FNV-1a hash to produce content addresses. FNV-1a is
//! fast, dependency-free, and fully deterministic across platforms and
//! languages, which makes it ideal for a build cache where the only
//! requirement is that identical bytes map to identical addresses.
//!
//! # Security note
//!
//! FNV-1a is **NOT** a cryptographic hash. It provides no collision or
//! pre-image resistance against a malicious adversary. Cairn is a build cache,
//! not a security boundary: do not rely on these digests to detect deliberate
//! tampering. See `docs/FORMAT.md` for the full rationale.

/// FNV-1a 64-bit offset basis.
const FNV_OFFSET_BASIS: u64 = 0xcbf2_9ce4_8422_2325;
/// FNV-1a 64-bit prime.
const FNV_PRIME: u64 = 0x0000_0100_0000_01b3;

/// Streaming FNV-1a 64-bit hasher.
///
/// Feed bytes with [`Hasher::update`] and read the final digest with
/// [`Hasher::finalize`] (numeric) or [`Hasher::finalize_hex`] (canonical
/// 16-character lowercase hex string).
#[derive(Clone, Copy, Debug)]
pub struct Hasher {
    state: u64,
}

impl Default for Hasher {
    fn default() -> Self {
        Self::new()
    }
}

impl Hasher {
    /// Create a fresh hasher seeded with the FNV offset basis.
    pub fn new() -> Self {
        Self {
            state: FNV_OFFSET_BASIS,
        }
    }

    /// Absorb a chunk of bytes into the running digest.
    pub fn update(&mut self, bytes: &[u8]) {
        let mut state = self.state;
        for &byte in bytes {
            state ^= u64::from(byte);
            state = state.wrapping_mul(FNV_PRIME);
        }
        self.state = state;
    }

    /// Return the raw 64-bit digest of everything absorbed so far.
    pub fn finalize(&self) -> u64 {
        self.state

File: go/internal/cache/hash.go
// Package cache implements the Cairn content-addressed cache format in Go.
//
// It is a faithful, standard-library-only port of the Rust engine: the same
// FNV-1a 64-bit hashing scheme, the same canonical cache-key stream, and the
// same manifest JSON layout. A manifest written by this package can be read
// and restored by the Rust `cairn` binary and vice versa.
//
// Security note: FNV-1a is NOT cryptographic. See docs/FORMAT.md.
package cache

import (
	"fmt"
	"io"
	"os"
)

const (
	fnvOffsetBasis uint64 = 0xcbf29ce484222325
	fnvPrime       uint64 = 0x00000100000001b3
)

// Hasher is a streaming FNV-1a 64-bit hasher, matching rust/src/hash.rs.
type Hasher struct {
	state uint64
}

// NewHasher returns a hasher seeded with the FNV offset basis.
func NewHasher() *Hasher {
	return &Hasher{state: fnvOffsetBasis}
}

// Update absorbs bytes into the running digest.
func (h *Hasher) Update(b []byte) {
	state := h.state
	for _, c := range b {
		state ^= uint64(c)
		state *= fnvPrime
	}
	h.state = state
}

// Sum returns the raw 64-bit digest.
func (h *Hasher) Sum() uint64 { return h.state }

// Hex returns the canonical 16-char lowercase hex digest.
func (h *Hasher) Hex() string { return fmt.Sprintf("%016x", h.state) }

// HashBytes hashes a byte slice and returns the canonical hex digest.
func HashBytes(b []byte) string {
	h := NewHasher()
	h.Update(b)
	return h.Hex()
}

// HashFile streams a file through the hasher, returning digest and size.
func HashFile(path string) (string, uint64, error) {
	f, err := os.Open(path)
	if err != nil {
		return "", 0, err
	}
	defer f.Close()

	h := NewHasher()
	buf := make([]byte, 64*1024)
	var total uint64
	for {
		n, err := f.Read(buf)
